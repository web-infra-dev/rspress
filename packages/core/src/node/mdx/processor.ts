import path from 'node:path';
import { createProcessor } from '@mdx-js/mdx';
import type { Header, UserConfig } from '@rspress/shared';
import { extractTextAndId, loadFrontMatter } from '@rspress/shared/node-utils';

import type { PluginDriver } from '../PluginDriver';
import type { RouteService } from '../route/RouteService';
import {
  applyReplaceRules,
  escapeMarkdownHeadingIds,
  normalizePath,
} from '../utils';
import { createMDXOptions } from './options';
import type { TocItem } from './remarkPlugins/toc';
import type { PageMeta } from './types';

interface CompileOptions {
  source: string;
  filepath: string;

  // assume that the below instances are singleton, it will not change.
  docDirectory: string;
  config: UserConfig | null;
  routeService: RouteService | null;
  pluginDriver: PluginDriver | null;
}

async function compile(options: CompileOptions): Promise<string> {
  const { source, filepath, docDirectory, config, routeService, pluginDriver } =
    options;

  const mdxOptions = await createMDXOptions({
    config,
    docDirectory,
    filepath,
    pluginDriver,
    routeService,
  });
  // Separate frontmatter and content in MDX source
  const { frontmatter, emptyLinesSource } = loadFrontMatter(
    source,
    filepath,
    docDirectory,
    true,
  );

  // For loader error stack, so we use the source with frontmatter @see https://github.com/web-infra-dev/rspress/issues/2010
  const { replaceRules } = config ?? {};
  const replacedSource = applyReplaceRules(emptyLinesSource, replaceRules);

  // Support custom id like `#hello world {#custom-id}`.
  // TODO > issue: `#hello world {#custom-id}` will also be replaced.
  const preprocessedContent = escapeMarkdownHeadingIds(replacedSource);

  try {
    let pageMeta: PageMeta = {
      title: '',
      toc: [] as TocItem[],
      headingTitle: '',
    };

    const frontmatterTitle = extractTextAndId(frontmatter.title)[0];

    const compiler = createProcessor(mdxOptions);

    compiler.data('pageMeta' as any, { toc: [], title: '' });
    const vFile = await compiler.process({
      value: preprocessedContent,
      path: filepath,
    });

    const compileResult = String(vFile);
    const compilationMeta = compiler.data('pageMeta' as any) as {
      toc: Header[];
      title: string;
    };

    const headingTitle = extractTextAndId(compilationMeta.title)[0];
    pageMeta = {
      ...compilationMeta,
      title: frontmatterTitle || headingTitle,
      headingTitle,
      frontmatter,
    } as PageMeta;

    const result = `const frontmatter = ${JSON.stringify(frontmatter)};
${compileResult}
MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["${encodeURIComponent(
      normalizePath(path.relative(docDirectory, filepath)),
    )}"] = ${JSON.stringify(pageMeta)};
`;
    return result;
  } catch (e: unknown) {
    if (e instanceof Error) {
      throw e;
    }
    throw new Error(`MDX compile error: ${e} in ${filepath}`);
  }
}

const cache = new Map<string, Promise<string>>();

async function compileWithCrossCompilerCache(
  options: CompileOptions,
): Promise<string> {
  const filepath = options.filepath;
  const task = cache.get(filepath);
  if (task) {
    // only for web and node, one write and one read
    // free the memory after one read
    cache.delete(filepath);
    return task;
  }

  const promise = compile(options);
  cache.set(filepath, promise);
  return promise;
}

export { compile, compileWithCrossCompilerCache };
