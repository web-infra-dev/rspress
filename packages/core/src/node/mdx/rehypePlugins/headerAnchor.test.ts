import { describe, it } from 'vitest';
import { compile } from '../processor';

describe('rehypeHeadAnchor', () => {
  it('basic', async () => {
    const result = await compile({
      source: `
# Guide

## title 1 {#custom-id}

## title 1 \{#custom-id}

## title 2

## title 2

## \`title\` 2

## Title 2

## **title** 2

{/* prettier-ignore-start */}
## *title* 2
{/* prettier-ignore-end */}

`,
      checkDeadLinks: false,
      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/index.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });
    expect(result).toMatchSnapshot();
  });

  it('should render inline code in title', async () => {
    const result = await compile({
      source: `
# Hello World \`inline code\`
`,
      checkDeadLinks: false,
      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/inline-code.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });
    expect(result).toMatchSnapshot();
  });

  it('should support custom id', async () => {
    const result = await compile({
      source: `
# Hello World \`inline code\` \\{#custom-id}
`,
      checkDeadLinks: false,
      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/inline-code.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });

    // This is actually a wrong usage, but we need to be compatible.
    const resultWithoutBackslash = await compile({
      source: `
# Hello World \`inline code\` {#custom-id}
`,
      checkDeadLinks: false,
      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/inline-code.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });
    expect(resultWithoutBackslash).toEqual(result);
    expect(result).toMatchSnapshot();
  });
});
