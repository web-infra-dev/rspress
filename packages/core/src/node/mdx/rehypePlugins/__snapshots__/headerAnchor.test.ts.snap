// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rehypeHeadAnchor > basic 1`] = `
"const frontmatter = {};
/*prettier-ignore-start*/
/*prettier-ignore-end*/
import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    code: "code",
    em: "em",
    h1: "h1",
    h2: "h2",
    strong: "strong",
    ..._provideComponents(),
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsxs(_components.h1, {
      id: "guide",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#guide",
        children: "#"
      }), "Guide"]
    }), "\\n", _jsxs(_components.h2, {
      id: "custom-id",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#custom-id",
        children: "#"
      }), "title 1"]
    }), "\\n", _jsxs(_components.h2, {
      id: "custom-id",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#custom-id",
        children: "#"
      }), "title 1"]
    }), "\\n", _jsxs(_components.h2, {
      id: "title-2",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#title-2",
        children: "#"
      }), "title 2"]
    }), "\\n", _jsxs(_components.h2, {
      id: "title-2-1",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#title-2-1",
        children: "#"
      }), "title 2"]
    }), "\\n", _jsxs(_components.h2, {
      id: "title-2-2",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#title-2-2",
        children: "#"
      }), _jsx(_components.code, {
        children: "title"
      }), " 2"]
    }), "\\n", _jsxs(_components.h2, {
      id: "title-2-3",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#title-2-3",
        children: "#"
      }), "Title 2"]
    }), "\\n", _jsxs(_components.h2, {
      id: "title-2-4",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#title-2-4",
        children: "#"
      }), _jsx(_components.strong, {
        children: "title"
      }), " 2"]
    }), "\\n", "\\n", _jsxs(_components.h2, {
      id: "title-2-5",
      children: [_jsx(_components.a, {
        className: "header-anchor",
        "aria-hidden": "true",
        href: "#title-2-5",
        children: "#"
      }), _jsx(_components.em, {
        children: "title"
      }), " 2"]
    }), "\\n"]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["index.mdx"] = {"toc":[{"id":"custom-id","text":"title 1","depth":2},{"id":"custom-id","text":"title 1","depth":2},{"id":"title-2","text":"title 2","depth":2},{"id":"title-2-1","text":"title 2","depth":2},{"id":"title-2-2","text":"\`title\` 2","depth":2},{"id":"title-2-3","text":"Title 2","depth":2},{"id":"title-2-4","text":"**title** 2","depth":2},{"id":"title-2-5","text":"*title* 2","depth":2}],"title":"Guide","headingTitle":"Guide","frontmatter":{}};
"
`;

exports[`rehypeHeadAnchor > should render inline code in title 1`] = `
"const frontmatter = {};
import {jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    code: "code",
    h1: "h1",
    ..._provideComponents(),
    ...props.components
  };
  return _jsxs(_components.h1, {
    id: "hello-world-inline-code",
    children: [_jsx(_components.a, {
      className: "header-anchor",
      "aria-hidden": "true",
      href: "#hello-world-inline-code",
      children: "#"
    }), "Hello World ", _jsx(_components.code, {
      children: "inline code"
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["inline-code.mdx"] = {"toc":[],"title":"Hello World \`inline code\`","headingTitle":"Hello World \`inline code\`","frontmatter":{}};
"
`;

exports[`rehypeHeadAnchor > should support custom id 1`] = `
"const frontmatter = {};
import {jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    code: "code",
    h1: "h1",
    ..._provideComponents(),
    ...props.components
  };
  return _jsxs(_components.h1, {
    id: "custom-id",
    children: [_jsx(_components.a, {
      className: "header-anchor",
      "aria-hidden": "true",
      href: "#custom-id",
      children: "#"
    }), "Hello World ", _jsx(_components.code, {
      children: "inline code"
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["inline-code.mdx"] = {"toc":[],"title":"Hello World \`inline code\`","headingTitle":"Hello World \`inline code\`","frontmatter":{}};
"
`;

exports[`rehypeHeadAnchor > should support mdx component with trim 1`] = `
"const frontmatter = {};
import {jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    h1: "h1",
    ..._provideComponents(),
    ...props.components
  }, {Badge} = _components;
  if (!Badge) _missingMdxReference("Badge", true);
  return _jsxs(_components.h1, {
    id: "hello-world",
    children: [_jsx(_components.a, {
      className: "header-anchor",
      "aria-hidden": "true",
      href: "#hello-world",
      children: "#"
    }), "Hello World ", _jsx(Badge, {
      text: "WARNING"
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
function _missingMdxReference(id, component) {
  throw new Error("Expected " + (component ? "component" : "object") + " \`" + id + "\` to be defined: you likely forgot to import, pass, or provide it.");
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["inline-code.mdx"] = {"toc":[],"title":"Hello World ","headingTitle":"Hello World ","frontmatter":{}};
"
`;
